src/v/cluster/leader_router.h

```diff
@@ -0,0 +1,266 @@
+concept IsHandler = requires(req_t req, resp_t resp, handler_t handler) {
+    // RPC protocol with which to perform remote dispatch.
+    { typename handler_t::proto_t(nullptr) };
+
+    // Returns a short description of the process being dispatched to leader,
+    // suitable for logging.
+    { handler_t::process_name() } -> std::same_as<ss::sstring>;
+
+    // Returns an error response with the given code.
+    { handler_t::error_resp(cluster::errc()) } -> std::same_as<resp_t>;
+
+    // Sends the given request on the given client protocol with the given
+    // timeout.
+    {
+        handler_t::dispatch(
+          typename handler_t::proto_t(nullptr),
+          req,
+          model::timeout_clock::duration())
+    } -> std::same_as<ss::future<result<rpc::client_context<resp_t>>>>;
+
+    // Processes the given request on the given shard. This is expected to run
+    // on the leader.
+    {
+        handler.process(ss::shard_id(), req)
+    } -> std::same_as<ss::future<resp_t>>;
+};
+
+// Encapsulates logic to direct RPCs to a specific NTP's leader.
+//
+// When a caller invokes process_or_dispatch() directed at an NTP whose leader
+// is on a different broker, the router redirects the request to the service
+// located on the leader broker.
+//
+// When a caller is colocated with the leader, the router calls
+// find_shard_and_process() directly, calling into the handler's process()
+// implementations.
+template<typename req_t, typename resp_t, typename handler_t>
+requires IsHandler<req_t, resp_t, handler_t>
+class leader_router {
+public:
+    using duration = model::timeout_clock::duration;
+    leader_router(
+      ss::sharded<cluster::shard_table>&,
+      ss::sharded<cluster::metadata_cache>&,
+      ss::sharded<rpc::connection_cache>&,
+      ss::sharded<partition_leaders_table>&,
+      handler_t&,
+      model::node_id self,
+      int16_t retries,
+      std::chrono::milliseconds timeout);
+
+    // Sends the given request to the leader, or processes it on the
+    // appropriate shard if this node is the leader.
+    ss::future<resp_t>
+    process_or_dispatch(req_t req, model::ntp, duration timeout);
+
+    // Finds the shard appropriate for the NTP and processes the given request,
+    // expecting the leader is on this node.
+    ss::future<resp_t>
+    find_shard_and_process(req_t req, model::ntp, duration timeout);
+
+    ss::future<> shutdown() {
+        _as.request_abort();
+        co_await _gate.close();
+    }
+
+private:
+    // Sends the given request to the given leader node ID..
+    virtual ss::future<resp_t>
+    dispatch_to_leader(req_t req, model::node_id, duration timeout);
+
+    ss::abort_source _as;
+    ss::gate _gate;
+    ss::sharded<cluster::shard_table>& _shard_table;
+    ss::sharded<cluster::metadata_cache>& _metadata_cache;
+    ss::sharded<rpc::connection_cache>& _connection_cache;
+    ss::sharded<partition_leaders_table>& _leaders;
+    handler_t& _handler;
```

Comment:
that the handler here is a reference has me wondering how all this is wired up and if `_handler` will always point to a handler instance on "this" core...

i saw where it was initialized in the constructor for this
```
class allocate_id_router
  : public leader_router<
      allocate_id_request,
      allocate_id_reply,
      allocate_id_handler> {

private:
    allocate_id_handler _handler;
};
```

so then it's embedded a member in `id_allocator_frontend` so then presumably is all part of the vertical `seastar::sharded<id_allocator_frontend>` stack?

Reply:
Yeah, the `leader_router` and `handler` are currently expected to be tightly coupled and on the same shard. It's not the prettiest, but the idea is that the handler and router can be initialized together as a sharded service if need-be, but the requirement is that the leader router must be able to handle dispatch to the local node.

I'd considered making the handler sharded here, but found it a bit more unwieldy to first have to construct the sharded handler elsewhere (likely in redpanda/application). Tying the router and handler allows both to be constructed together, though I'm open to suggestions though.

Reply:
Was wondering if this can do this without passing an explicit handler and use lambdas instead, something like `local_router<DispatchFunc, LocalFunc>{args}.invoke()`  .. think that simplifies things further

1. we don't need explicit req_t, resp_t, we can deduce them from the return type of lambdas
2. we don't need inheritance and we don't need to pass around the args (ntps, timeout and such)
3. the handler to shard association is probably more obvious for a reader as they see a lambda on the caller side.

Reply:
>Was wondering if we can...

Such a method would need to be templated on resp_t since it'd be a part of the return type, and it'd have to know how to create an erroneous resp_t when it isn't the leader. I'm also not sure how this would look since _something_ would need to know about the req_t in order to pass the bytes around (though maybe you're saying each func should have a copy of the request).

All in all, maybe we can. The vision isn't clear in my head, and I'd rather not spend more time on this if you don't feel strongly about it.

It's a good callout that the router-handler relationship isn't obvious. Will ponder ways to make it more obvious. Really, I want the handler to be owned by the router. I'm not sure how to structure the construction of it other than subclassing, since each handler may have different constructor args.

Reply:
> Such a method would need to be templated on resp_t since it'd be a part of the return type, and it'd have to know how to create an erroneous resp_t when it isn't the leader. I'm also not sure how this would look since something would need to know about the req_t in order to pass the bytes around (though maybe you're saying each func should have a copy of the request).


I was thinking something like this..

    using func_result_t = std::invoke_result_t<DispatchFunc>::type; // future<allocate_id_reply>
    using underlying_t = seastar::is_future<func_result_t>::value; // allocate_id_reply
    using result_t = ss::future<result<underlying_t>>; // future<result<allocate_id_reply>>

result_t would propagate any error that the caller can later transform into its own return type. Think passing req_t is probably not as much of a problem.

> All in all, maybe we can. The vision isn't clear in my head, and I'd rather not spend more time on this if you don't feel strongly about it.

I don't feel strongly at all, was just throwing ideas for the sake of discussion (should've probably mentioned that in my original comment). The code as-is lgtm.

Reply:
We chatted about this offline. I think this would be a reasonable evolution of this, but I don't feel strongly enough to go down this route right now.